pkg/pillar/zedcloud/authen.go

```diff
@@ -0,0 +1,508 @@
+	if useOnboard {
+		sm.SenderCertHash = onBoardCertHash
+		sm.SenderCert = onBoardCertBytes
+		if len(sm.SenderCert) == 0 {
+			err := fmt.Errorf("addAuthentication: SenderCert empty")
+			log.Infof("addAuthenticate: get sender cert failed, %v\n", err)
+			return nil, err
+		}
+		log.Infof("addAuthenticate: onboard senderCert size %d\n", len(sm.SenderCert))
+	} else {
+		sm.SenderCertHash = deviceCertHash
+	}
+	if sm.SenderCertHash == nil {
+		err := fmt.Errorf("addAuthentication: SenderCertHash empty")
+		return nil, err
+	}
+
+	sig, err := signAuthData(data, cert)
+	if err != nil {
+		log.Debugf("addAuthenticate: sign auth data error %v\n", err)
+		return nil, err
+	}
+	sm.SignatureHash = sig
+	sm.Algo = zauth.HashAlgorithm_HASH_SHA256_16bytes
+
+	data2, err := proto.Marshal(&sm)
+	if err != nil {
+		log.Debugf("addAuthenticate: auth marshal error %v\n", err)
+		return nil, err
+	}
+
+	size := int64(proto.Size(&sm))
+	buf := bytes.NewBuffer(data2)
+	log.Debugf("addAuthenticate: payload size %d, sig %v\n", size, sig)
+	return buf, nil
+}
+
+func getMyDevCert(isOnboard bool) (tls.Certificate, error) {
+	var cert tls.Certificate
+	var err error
+	if isOnboard {
+		if onBoardCert == nil {
+			cert, err = tls.LoadX509KeyPair(types.OnboardCertName,
+				types.OnboardKeyName)
+			if err != nil {
+				log.Debugf("getMyDevCert: get onboard cert error %v\n", err)
+				return cert, err
+			}
+
+			onboardCertpem, err := ioutil.ReadFile(types.OnboardCertName)
+			if err != nil {
+				log.Debugf("getMyDevCert: get onboard certbytes error %v\n", err)
+				return cert, err
+			}
+			onBoardCertBytes = []byte(base64.StdEncoding.EncodeToString(onboardCertpem))
+
+			// device side of cert hash is calculated from the x509 cert, not from PEM bytes
+			onBoardCertHash, err = certToSha256(cert, true)
+			if err != nil {
+				return cert, err
+			}
+			onBoardCert = &cert
+			log.Infof("getMyDevCert: onboard cert with hash %v\n", string(onBoardCertHash))
+		} else {
+			cert = *onBoardCert
+		}
+	} else {
+		if deviceCert == nil {
+			cert, err = GetClientCert()
+			if err != nil {
+				log.Infof("getMyDevCert: get client cert error %v\n", err)
+				return cert, err
+			}
+
+			// device side of cert hash is calculated from the x509 cert, not from PEM bytes
+			deviceCertHash, err = certToSha256(cert, false)
+			if err != nil {
+				return cert, err
+			}
+			deviceCert = &cert
+			log.Infof("getMyDevCert: device cert with hash %v\n", string(deviceCertHash))
+		} else {
+			cert = *deviceCert
+		}
+	}
+	return cert, nil
+}
+
+// sign hash'ed data with certificate private key
+func signAuthData(sigdata []byte, cert tls.Certificate) ([]byte, error) {
+	log.Debugf("sending sigdata (len %d) %v\n", len(sigdata), sigdata)
+	hash := ComputeSha(sigdata)
+
+	var sigres []byte
+	switch key := cert.PrivateKey.(type) {
+	default:
+		//log.Fatal("Private Key RSA type not supported")
+		err := fmt.Errorf("signAuthData: privatekey default")
```

Comment:
Please include %T of the type in the error string.
Any reason to change from log.Fatal?

Reply:
OK.